### PR TITLE
feat: apply retention delete predicate

### PR DESCRIPTION
**Pull request by NGA-TRAN**
Closes https://github.com/influxdata/influxdb_iox/issues/6028
This is the second and last PR for https://github.com/influxdata/influxdb_iox/issues/6028 to filter rows on remaining chunks

Note that this PR on to https://github.com/influxdata/influxdb_iox/pull/6197 (not master) 

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: querier/src/table/mod.rs [hunk 1]
```diff
@@ -1241,6 +1258,46 @@ mod tests {
...
+            .await;
+
+        let max_parquet_sequence_number = Some(SequenceNumber::new(1));
+        let max_tombstone_sequence_number = Some(SequenceNumber::new(2));
+        let ingester_partition =
+            builder.build(max_parquet_sequence_number, max_tombstone_sequence_number);
+
+        let querier_table = querier_table.with_ingester_partition(ingester_partition);
+
+        // There must be two delete predicates: one from the tombstone, one from retention
```
Comment (NGA-TRAN) [inline querier/src/table/mod.rs:1296]: This is the purpose of this test to see that the  retention delete predicate is added for namespace with retention period

File: querier/src/table/mod.rs [hunk 2]
```diff
@@ -232,20 +234,32 @@ impl QuerierTable {
...
+                    let retention_time_ns = self
+                        .chunk_adapter
+                        .catalog_cache()
+                        .time_provider()
+                        .now()
+                        .timestamp_nanos()
+                        - retention_period.as_nanos() as i64;
+
+                    // Add predicate to only keep chunks inside the retention period: time >= retention_period
+                    let predicate = Some(predicate.clone().with_retention(retention_time_ns));
```
Comment (NGA-TRAN) [inline querier/src/table/mod.rs:250]: This predicate is used to early prune chunks that are fully outside the retention period. This was available on previous PR. This is just a refactor

File: querier/src/table/mod.rs [hunk 3]
```diff
@@ -232,20 +234,32 @@ impl QuerierTable {
...
+
+                    // Expression used to add to delete predicate to delete data older than retention period
+                    // time < retention_time
+                    let retention_delete_pred = Some(DeletePredicate::retention_delete_predicate(
```
Comment (NGA-TRAN) [inline querier/src/table/mod.rs:254]: This is the delete predicate to apply on remaining chunks which is the purpose of this PR

File: querier/src/namespace/query_access.rs [hunk 4]
```diff
@@ -685,21 +684,25 @@ mod tests {
...
+                "|               |   ProjectionExec: expr=[host@0 as host, load@1 as load, time@2 as time]                                                                 |",
+                "|               |     DeduplicateExec: [host@0 ASC,time@2 ASC]                                                                                            |",
+                "|               |       SortPreservingMergeExec: [host@0 ASC,time@2 ASC]                                                                                  |",
+                "|               |         UnionExec                                                                                                                       |",
+                "|               |           CoalesceBatchesExec: target_batch_size=4096                                                                                   |",
+                "|               |             FilterExec: time@2 < -9223372036854775808 OR time@2 > -3600000000000                                                        |",
```
Comment (NGA-TRAN) [inline querier/src/namespace/query_access.rs:699]: This the result of NOT(MIN < time < retention_time)

File: query_tests/src/scenarios/util.rs [hunk 5]
```diff
@@ -653,7 +653,9 @@ impl MockIngester {
     async fn new() -> Self {
         let exec = Arc::clone(&GLOBAL_EXEC);
         let catalog = TestCatalog::with_execs(exec, 4);
-        let ns = catalog.create_namespace_1hr_retention("test_db").await;
+        let ns = catalog
+            .create_namespace_with_retention("test_db", None)
+            .await;
```
Comment (NGA-TRAN) [inline query_tests/src/scenarios/util.rs:658]: Need to make this infinite to have the available tests do what they expect to test

**Comment by NGA-TRAN** [comment]
@alamb and @crepererum : to make it easy to review, I compare the new branch with the previous branch that is not merged yet (https://github.com/influxdata/influxdb_iox/pull/6197). I am not sure how to merge this into main without asking you to review again. Maybe I wait for you to finish reviewing both branches, then I merge this branch into the other branch and then merge the other branch into main

File: querier/src/namespace/query_access.rs [hunk 6]
```diff
@@ -671,7 +671,6 @@ mod tests {
                 "| host | load | time                 |",
                 "+------+------+----------------------+",
                 "| a    | 1    | 1970-01-01T00:00:00Z |",
-                "| aa   | 11   | 1969-12-31T22:00:00Z |", // todo: this line will go away after my next PR to apply predicate on partially outside retention chunks
```
Comment (alamb) [inline querier/src/namespace/query_access.rs:674]: 🎉

File: querier/src/namespace/query_access.rs (hunk 4 above)
Comment (alamb) [reply querier/src/namespace/query_access.rs:699]: Eventually we should probably push this filtering down to the `ParquetExec` as well (so RowGroup / late materialization can be applied to), but perhaps we can track that as a follow on item in a PR

**Review by alamb — APPROVED** [review]

File: querier/src/namespace/query_access.rs (hunk 4 above)
Comment (NGA-TRAN) [reply querier/src/namespace/query_access.rs:699]: I will open a ticket for this

Comment (NGA-TRAN) [reply querier/src/namespace/query_access.rs:699]: #6210

**Pull request merged**